sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala

```diff
@@ -347,8 +348,9 @@ case class AttributeReference(
   }
 
   override def sql: String = {
-    val qualifierPrefix = if (qualifier.nonEmpty) qualifier.mkString(".") + "." else ""
-    s"$qualifierPrefix${quoteIdentifier(name)}"
+    val qualifierPrefix =
+      if (qualifier.nonEmpty) qualifier.map(quoteIfNeeded).mkString(".") + "." else ""
+    s"$qualifierPrefix${quoteIfNeeded(name)}"
```

Comment:
@sarutak @cloud-fan previously we always add backticks. But now we only add backticks when `name` contains dot or backtick. Is there any other valid case that may require backticks even if they don't contain dot or backtick, such as spaces in `name`? For example,

```
spark.sessionState.sqlParser.parseExpression("`a b` + 1").sql
```
It returns the following string before the change:
```
(`a b` + 1)
```
But now it returns
```
(a b + 1)
```

Reply:
The first idea of mine was always adding backticks but finally I used `quoteIfNeeded` after discussion.
We might need to quote none of alphabets, numbers, underscore.
Or, do you think it's better to always quote ?

Reply:
This is a good point. For places that only display the table/column name, like in the error message `Table not found: a b`, it's a bit weird but users can still understand the meaning. But in an expression sql string, `a b + 1` is ambiguous.

I think we shoud follow the parser definition of identifiers `(LETTER | DIGIT | '_')+`, and only omit the quoting if the name only contains letter, digit or underscore.

Reply:
O.K, I'll open a PR to fix it.